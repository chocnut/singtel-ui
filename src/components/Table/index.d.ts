declare const __DEV__: boolean;
declare const __VERSION__: string;
declare const $FixMe: any;
