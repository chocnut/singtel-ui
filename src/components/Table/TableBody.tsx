import React, { FC, useState, useCallback } from 'react';
import useIsMobile from './hooks/useIsMobile';
import {
  InputRadio,
  SelectionContainer,
  TableBodyContainer,
  TableBodyContent,
  TableRow,
} from './styles';

export interface TableBodyProps {
  columns: [];
  dataSource: [];
  rowSelection?: {
    type: 'checkbox' | 'radio';
  };
}

const TableBody: FC<TableBodyProps> = ({
  columns,
  dataSource,
  rowSelection,
}) => {
  const [selectedRowKeys, setSelectedRowKeys] = useState([]);
  const isMobile = useIsMobile();

  const handleCheckBoxClick = (value: string) => {
    let newCheckboxArr = [...selectedRowKeys];
    if (newCheckboxArr.indexOf(value) >= 0) {
      newCheckboxArr.splice(newCheckboxArr.indexOf(value), 1);
    } else {
      newCheckboxArr.push(value);
    }
    setSelectedRowKeys(newCheckboxArr);
  };

  const handleRadioClick = (value: string) => {
    setSelectedRowKeys([value]);
  };

  const handleClick = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const value = e.target.value;

      rowSelection?.type === 'checkbox'
        ? handleCheckBoxClick(value)
        : handleRadioClick(value);
    },
    [selectedRowKeys]
  );

  const renderMobileRow = columnItem => {
    return (
      <TableRow mobile>
        {columns.map((item, index) => {
          return (
            <span key={index}>
              {item.title}: {columnItem[item.key]}
            </span>
          );
        })}
      </TableRow>
    );
  };

  return (
    <div>
      {dataSource?.map((columnItem, index) => {
        return (
          <TableBodyContainer
            key={index}
            selected={selectedRowKeys.includes(index.toString())}
          >
            <TableBodyContent>
              {rowSelection && (
                <SelectionContainer
                  type={rowSelection.type}
                  selected={selectedRowKeys.includes(index.toString())}
                >
                  <InputRadio
                    onClick={handleClick}
                    type={rowSelection.type}
                    name={rowSelection.type}
                    value={index}
                  />
                </SelectionContainer>
              )}
              {isMobile
                ? renderMobileRow(columnItem)
                : columns.map((item, index) => {
                    return (
                      <TableRow key={item?.key + index}>
                        {columnItem[item?.key]}
                      </TableRow>
                    );
                  })}
            </TableBodyContent>
          </TableBodyContainer>
        );
      })}
    </div>
  );
};

export default TableBody;
