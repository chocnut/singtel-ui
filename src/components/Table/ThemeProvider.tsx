import React, { FC, ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

export interface ThemeProviderProps {
  colors: {
    primary: string;
    light: string;
  };
  fontFamily: string;
}

export const mainTheme: ThemeProviderProps = {
  colors: {
    primary: '#5C50BB',
    light: '#EFEDFD',
  },
  fontFamily: 'Avenir LT Std, sans-serif, Roboto',
};

interface ThemeProps {
  children: ReactNode;
  theme?: object;
}
const Theme: FC<ThemeProps> = ({ children, theme }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

export default Theme;
