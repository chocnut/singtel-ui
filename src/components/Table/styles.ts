import styled, { css } from 'styled-components';

type SelectionContainer = {
  type: 'checkbox' | 'radio';
  selected?: boolean;
};

const StyledTable = styled.div`
  width: 100%;
  box-shadow: 0 3px 8px rgba(0, 0, 0, 0.12);
  border-radius: 16px;
  overflow: hidden;
`;

const TableHeaderContainer = styled.div`
  display: flex;
  align-items: center;
  color: #000;
  height: 80px;
  line-height: 80px;
  background: #f7f7f7;
`;

const TableHeaderContent = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  height: inherit;
  margin: 0 24px;
`;

const TableHeaderItem = styled.div`
  display: flex;
  flex-direction: row;
  flex: 1;
  align-items: center;
`;

const TableHeaderItemText = styled.div`
  font-weight: 700;
  font-size: 24px;
  font-style: normal;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const TableBodyContainer = styled.div`
  display: flex;
  align-items: center;
  color: #000;
  height: 84px;
  background: ${({ selected, theme }: { selected: boolean; theme: any }) =>
    selected ? theme.colors.light : ''};

  :hover {
    background: ${props => props.theme.colors.light};
  }

  @media screen and (max-width: 768px) {
    height: 100%;
  }
`;

const TableHeaderEmptyCell = styled.div`
  position: relative;
  margin-right: 24px;
  width: 32px;
  height: 32px;
`;

const TableBodyContent = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  height: inherit;
  color: #000;
  border-bottom: 1px solid #e1e1e1;
  margin: 0 24px;
`;

const TableRow = styled.div`
  flex: 1;
  text-align: initial;
  font-weight: 350;
  font-size: 20px;
  line-height: 35px;
  letter-spacing: 0.001em;
  font-style: normal;
  white-space: nowrap;
  width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;

  ${mobile =>
    mobile &&
    css`
      display: flex;
      flex-direction: column;
      padding: 10px 0 10px;

      .title {
        font-style: normal;
        font-weight: 700;
      }
    `}
`;

const SelectionContainer = styled.label`
  cursor: pointer;
  position: relative;
  margin-right: 24px;
  width: 32px;
  height: 32px;

  @media screen and (max-width: 768px) {
    top: -35px;
  }

  ${({ type, selected }: SelectionContainer) =>
    type === 'radio'
      ? selected
        ? css`
            background: ${props => props.theme.colors.primary};
            border-radius: 100%;
            border: 1px solid ${props => props.theme.colors.primary};

            ::before {
              content: '';
              position: absolute;
              left: 10px;
              width: 16px;
              height: 16px;
              background-color: #fff;
              border-radius: 100%;
              transform: translateX(-10%) translateY(50%);
            }
          `
        : css`
            border-radius: 100%;
            border: 1px solid #a8a8a8;
          `
      : selected
      ? css`
          background: ${props => props.theme.colors.primary};
          border-radius: 8px;
          border: 1px solid ${props => props.theme.colors.primary};

          ::before {
            content: '';
            position: absolute;
            left: 10px;
            width: 6px;
            height: 14px;
            border-color: #fff;
            border-style: solid;
            border-width: 0 2px 2px 0;
            transform: translateX(40%) translateY(35%) rotate(45deg);
          }
        `
      : css`
          border: 1px solid #a8a8a8;
          border-radius: 8px;
        `}
`;

const InputRadio = styled.input`
  opacity: 0;
  visibility: hidden;
  width: 0;
  height: 0;
  position: absolute;
  top: 0;
  left: 0;
`;

const SortingButton = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 52.34px;
  height: 56px;
  margin-left: 5px;
  border-radius: 100%;
  cursor: pointer;

  :hover {
    background: #dedafa;
  }
`;

export {
  StyledTable,
  TableHeaderContainer,
  TableHeaderContent,
  TableHeaderItem,
  TableHeaderItemText,
  TableHeaderEmptyCell,
  TableBodyContainer,
  TableBodyContent,
  TableRow,
  InputRadio,
  SelectionContainer,
  SortingButton,
};
