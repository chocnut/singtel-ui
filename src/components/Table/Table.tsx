import React, { FC, useState } from 'react';
import ThemeProvider, { mainTheme, ThemeProviderProps } from './ThemeProvider';
import { StyledTable } from './styles';
import TableBody from './TableBody';
import TableHeader from './TableHeader';

export interface Column {
  title: string;
  key: string;
  dataIndex: string;
  sorter?: (a: object, b: object) => number;
}

export interface TableProps {
  columns?: Array<Column>;
  dataSource?: object[];
  rowSelection?: {
    type: 'checkbox' | 'radio';
    sorter?: (a: object, b: object) => number;
  };
  theme?: ThemeProviderProps;
}

const Table: FC<TableProps> = ({
  columns = [],
  dataSource = [],
  rowSelection,
  theme = mainTheme,
}) => {
  const [data, setData] = useState<object[]>();
  const [order, setOrder] = useState(false);
  const [sortTitle, setSortTitle] = useState<string>();

  const showSelect = typeof rowSelection?.type === 'string';

  const onHandleSort = (item: Column) => {
    const sortedDataSource = dataSource?.sort(item?.sorter);
    setData(sortedDataSource);
    setOrder(!order);
    setSortTitle(item.title);
  };

  return (
    <ThemeProvider theme={theme}>
      <StyledTable>
        <TableHeader
          columns={columns}
          showSelect={showSelect}
          onHandleSort={onHandleSort}
          order={order}
          sortTitle={sortTitle}
        />
        <TableBody
          columns={columns}
          dataSource={data || dataSource}
          rowSelection={rowSelection}
        />
      </StyledTable>
    </ThemeProvider>
  );
};

export default Table;
