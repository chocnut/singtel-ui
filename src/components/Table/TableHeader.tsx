import React, { FC } from 'react';
import useIsMobile from './hooks/useIsMobile';
import SortArrowIcon from './icons/SortArrowIcon';
import SortDefaultIcon from './icons/SortDefaultIcon';
import {
  SortingButton,
  TableHeaderContainer,
  TableHeaderContent,
  TableHeaderEmptyCell,
  TableHeaderItem,
  TableHeaderItemText,
} from './styles';
import { Column } from './Table';

export interface TableHeaderProps {
  columns: Array<Column>;
  dataSource?: object[];
  onHandleSort?: (
    e: React.MouseEvent<HTMLButtonElement>,
    sorter: () => void
  ) => void;
  rowSelection: {
    type: 'checkbox' | 'radio';
  };
  order: boolean;
  sortTitle?: string;
}

const TableHeader: FC<TableHeaderProps> = ({
  columns,
  rowSelection,
  onHandleSort,
  order,
  sortTitle,
}) => {
  const isMobile = useIsMobile();

  const renderSortButton = item => {
    return (
      <SortingButton onClick={e => onHandleSort?.(e, item)}>
        {sortTitle && sortTitle === item.title ? (
          <SortArrowIcon
            style={{ transform: order ? 'rotate(-180deg)' : 'none' }}
          />
        ) : (
          <SortDefaultIcon />
        )}
      </SortingButton>
    );
  };

  return (
    <TableHeaderContainer>
      <TableHeaderContent>
        {rowSelection && <TableHeaderEmptyCell />}
        {isMobile ? (
          <TableHeaderItem>
            <TableHeaderItemText />
          </TableHeaderItem>
        ) : (
          columns?.map((item: any, index: number) => {
            return (
              <TableHeaderItem key={index}>
                <TableHeaderItemText key={index}>
                  {item.title}
                </TableHeaderItemText>
                {item.sorter ? renderSortButton(item) : null}
              </TableHeaderItem>
            );
          })
        )}
      </TableHeaderContent>
    </TableHeaderContainer>
  );
};

export default TableHeader;
