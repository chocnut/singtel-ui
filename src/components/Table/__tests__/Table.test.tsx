import React from 'react';
import { render } from '@testing-library/react';

import Table from '../Table';

describe('Table', () => {
  const columns = [
    {
      title: 'Name',
      key: 'name',
      dataIndex: 'name',
    },
    {
      title: 'Mobile',
      key: 'mobile',
      dataIndex: 'mobile',
    },
    {
      title: 'Expiry',
      key: 'expiry',
      dataIndex: 'expiry',
    },
    {
      title: 'Penalty',
      key: 'penalty',
      dataIndex: 'penalty',
    },
  ];

  const dataSource = [
    {
      id: 1,
      name: 'Mavis Chen',
      mobile: '8899 7654',
      expiry: 'Dec 2022',
      penalty: '$600',
    },
    {
      id: 2,
      name: 'Rodney Artichoke',
      mobile: '9988 7676',
      expiry: 'Dec 2022',
      penalty: '$300',
    },
    {
      id: 3,
      name: 'Valentino Morose',
      mobile: '8900 7654',
      expiry: 'Dec 2022',
      penalty: '$300',
    },
    {
      id: 4,
      name: 'Eric Widget',
      mobile: '8900 7654',
      expiry: 'Dec 2022',
      penalty: '$300',
    },
  ];

  it('render the component', () => {
    const { asFragment } = render(
      <Table columns={columns} dataSource={dataSource} />
    );

    expect(asFragment().firstChild).toMatchSnapshot();
  });

  it('does not crash when dataSource is empty', () => {
    render(
      <Table
        columns={[
          {
            title: 'Name',
            key: 'name',
            dataIndex: 'name',
          },
        ]}
      />
    );
  });

  it('updates columns', () => {
    const columns = [
      {
        title: 'Name',
        key: 'name',
        dataIndex: 'name',
      },
    ];
    const { container, rerender } = render(<Table columns={columns} />);

    const newColumns = [
      {
        title: 'Title',
        key: 'title',
        dataIndex: 'title',
      },
    ];
    rerender(<Table columns={newColumns} />);
    expect(container.querySelector('div')?.textContent).toEqual('Title');
  });
});
