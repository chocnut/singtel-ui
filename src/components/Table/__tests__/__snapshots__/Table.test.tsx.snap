// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table render the component 1`] = `
<div
  class="sc-bcXHqe bnLSHI"
>
  <div
    class="sc-gswNZR biQHsU"
  >
    <div
      class="sc-dkrFOg ENPGY"
    >
      <div
        class="sc-hLBbgP cczCkk"
      >
        <div
          class="sc-eDvSVe jVTtnv"
        >
          Name
        </div>
      </div>
      <div
        class="sc-hLBbgP cczCkk"
      >
        <div
          class="sc-eDvSVe jVTtnv"
        >
          Mobile
        </div>
      </div>
      <div
        class="sc-hLBbgP cczCkk"
      >
        <div
          class="sc-eDvSVe jVTtnv"
        >
          Expiry
        </div>
      </div>
      <div
        class="sc-hLBbgP cczCkk"
      >
        <div
          class="sc-eDvSVe jVTtnv"
        >
          Penalty
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      class="sc-jSUZER bdBamB"
    >
      <div
        class="sc-iBYQkv eLEYUK"
      >
        <div
          class="sc-ftTHYK cpHedF"
        >
          Mavis Chen
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          8899 7654
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          Dec 2022
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          $600
        </div>
      </div>
    </div>
    <div
      class="sc-jSUZER bdBamB"
    >
      <div
        class="sc-iBYQkv eLEYUK"
      >
        <div
          class="sc-ftTHYK cpHedF"
        >
          Rodney Artichoke
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          9988 7676
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          Dec 2022
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          $300
        </div>
      </div>
    </div>
    <div
      class="sc-jSUZER bdBamB"
    >
      <div
        class="sc-iBYQkv eLEYUK"
      >
        <div
          class="sc-ftTHYK cpHedF"
        >
          Valentino Morose
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          8900 7654
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          Dec 2022
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          $300
        </div>
      </div>
    </div>
    <div
      class="sc-jSUZER bdBamB"
    >
      <div
        class="sc-iBYQkv eLEYUK"
      >
        <div
          class="sc-ftTHYK cpHedF"
        >
          Eric Widget
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          8900 7654
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          Dec 2022
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          $300
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table renders 1`] = `
<div
  class="sc-bcXHqe bnLSHI"
>
  <div
    class="sc-gswNZR biQHsU"
  >
    <div
      class="sc-dkrFOg ENPGY"
    >
      <div
        class="sc-hLBbgP cczCkk"
      >
        <div
          class="sc-eDvSVe jVTtnv"
        >
          Name
        </div>
      </div>
      <div
        class="sc-hLBbgP cczCkk"
      >
        <div
          class="sc-eDvSVe jVTtnv"
        >
          Mobile
        </div>
      </div>
      <div
        class="sc-hLBbgP cczCkk"
      >
        <div
          class="sc-eDvSVe jVTtnv"
        >
          Expiry
        </div>
      </div>
      <div
        class="sc-hLBbgP cczCkk"
      >
        <div
          class="sc-eDvSVe jVTtnv"
        >
          Penalty
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      class="sc-jSUZER bdBamB"
    >
      <div
        class="sc-iBYQkv eLEYUK"
      >
        <div
          class="sc-ftTHYK cpHedF"
        >
          Mavis Chen
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          8899 7654
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          Dec 2022
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          $600
        </div>
      </div>
    </div>
    <div
      class="sc-jSUZER bdBamB"
    >
      <div
        class="sc-iBYQkv eLEYUK"
      >
        <div
          class="sc-ftTHYK cpHedF"
        >
          Rodney Artichoke
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          9988 7676
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          Dec 2022
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          $300
        </div>
      </div>
    </div>
    <div
      class="sc-jSUZER bdBamB"
    >
      <div
        class="sc-iBYQkv eLEYUK"
      >
        <div
          class="sc-ftTHYK cpHedF"
        >
          Valentino Morose
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          8900 7654
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          Dec 2022
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          $300
        </div>
      </div>
    </div>
    <div
      class="sc-jSUZER bdBamB"
    >
      <div
        class="sc-iBYQkv eLEYUK"
      >
        <div
          class="sc-ftTHYK cpHedF"
        >
          Eric Widget
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          8900 7654
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          Dec 2022
        </div>
        <div
          class="sc-ftTHYK cpHedF"
        >
          $300
        </div>
      </div>
    </div>
  </div>
</div>
`;
